Source/Initialization/PlasmaInjector.cpp

```diff
@@ -105,9 +106,61 @@ PlasmaInjector::PlasmaInjector (int ispecies, const std::string& name)
         mass = species::get_mass( physical_species );
     }
 
+    std::string s_inj_style;
+    pp.query("injection_style", s_inj_style);
+
     // parse charge and mass
     std::string charge_s;
+    std::string mass_s;
     bool charge_is_specified = pp.query("charge", charge_s);
+    bool mass_is_specified = pp.query("mass", mass_s);
+
+    if (s_inj_style == "external_file") {
+        external_file = true;
+        std::string str_injection_file;
+        pp.get("injection_file", str_injection_file);
+
+#ifdef WARPX_USE_OPENPMD
+        if (ParallelDescriptor::IOProcessor()) {
+            m_openpmd_input_series = std::make_unique<openPMD::Series>(
+                str_injection_file, openPMD::AccessType::READ_ONLY);
+
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                m_openpmd_input_series->iterations.size() == 1u,
+                "External file should contain only 1 iteration\n");
+            openPMD::Iteration it = m_openpmd_input_series->iterations.begin()->second;
+            AMREX_ALWAYS_ASSERT_WITH_MESSAGE(
+                it.particles.size() == 1u,
+                "External file should contain only 1 species\n");
+            openPMD::ParticleSpecies ps = it.particles.begin()->second;
+
+            // TODO: Add ASSERT_WITH_MESSAGE to test if mass and charge are constant records
+            if (!charge_is_specified) {
+                amrex::ParticleReal const p_q = ps["charge"][openPMD::RecordComponent::SCALAR].loadChunk<amrex::ParticleReal>().get()[0];
+                double const charge_unit = ps["charge"][openPMD::RecordComponent::SCALAR].unitSI();
```

Comment:
Thanks and good point: the base standard does not require this. So we should check and go on, because our users can still set the charge via the input file option. Will update.